Technotes/001-async-networking.md

```diff
@@ -0,0 +1,246 @@
+
+## Notable Changes
+
+### The Collection of Active and Pending Tasks
+
+This proposal includes a change to the collection of active / pending tasks used, in part, to support the `ImageFetcher.cancel(_:)` method that deserves additional discussion:
+
+- Current: `var tasks: Set<ImageFetcherTask>`
+- Proposed: `var tasks: Dictionary<String, Task<ImageSource, Error>>`
+
+#### Current `tasks` Implementation
+
+In the current implementation each call to `ImageFetcher.task(_:)` and `.load(_:)` creates a new task if the requested resource is not cached. That means that if two different sections of code request the same resource and one subsequently cancels its request, the other is unaffected. However, that also means the caller is responsible for preventing duplicate requests in cases where that is likely to be a concern.
+
+The `ImageFetcherTask` provides access to the `configuration`, `handler`, and `result` of a given operation. While it seems like it should be very unlikely, this present a potential thread-safety issue as it is possible for the caller to try reading a task’s `result` on one thread while the underlying `Operation`’s `completionBlock` mutates the task’s `result` on another thread or for the caller to mutate its `configuration` or `handler` member while it is being read on another thread.
+
+#### Proposed `tasks` Implementation
+
+This proposal changes the `tasks` collection to a `Dictionary` and first checks for an active/pending for the requested configuration before checking for a cached image or requesting one. This absolves the caller of responsibility for preventing duplicate requests, but it also means that any call to cancel an active / pending request cancels it for all callers.
+
+I replaced `ImageFetcherTask` by providing direct access to the underlying `Task`, which offers access to the request’s eventual value and a `cancel()` method. The result is more restricted but safer since we remove possibility of concurrently reading/writing task members.
+
+`ImageFetcher` will use an `NSLock` to ensure thread-safe access to `tasks`. I considered making it an `actor` but that seems like overkill when we only need to protect `tasks` and all access to it will be very quick.
+
+### Managing Concurrency
+
+This proposal also removes the ability to limit concurrent requests and cache lookups via `maxConcurrent` by making the cooperative thread pool responsible for managing those operations. Instead, `maxConcurrent` is limited to controlling concurrent processing of image data ultimately performed by the `Image.decompress(...)` method. I reasoned that this represents the most expensive operation performed by `ImageFetcher` and the one the OS is least suited to manage. [Previous efforts](https://github.com/Mobelux/ImageFetcher/pull/6#issuecomment-1115291685) at limiting concurrency outside of an `OpertionQueue` were not especially fruitful.
```

Comment:
I agree that letting the cooperative thread pool manage threads is the way to go. 
What priority do the tasks execute at? Are there any considerations for memory usage?

Reply:
I was thinking we could add a `priority` member to `ImageConfiguration`, though we could also add a `priority` argument to the relevant `ImageFetcher` members.

Re: memory usage, are you thinking of usage in connection with image processing specifically? I suppose `maxConcurrent` of the `ImageProcessor` would allow you to exert some control over that, but no, I hadn't planned on anything beyond that.